qt6: new packages

wdconinc opened the pull request:
This is an initial set of qt6 packages. Start of a discussion to get these included. @sethrj

## Why separate packages when qt4 and qt5 were a single package?
An attempt at having more flexibility for downstream package developers who may just need the base package, without e.g. and opengl or QtQuick.

wdconinc committed:
qt6: initial commit of several basic qt6 packages

sethrj commented on var/spack/repos/builtin/packages/qt6declarative/package.py:
```diff
@@ -0,0 +1,49 @@
+
+    versions = ['6.2.3']
+    for v in versions:
+        depends_on('qt6base@{}'.format(v), when='@{}'.format(v))
+        depends_on('qt6shadertools@{}'.format(v), when='@{}'.format(v))
```
```suggestion
        depends_on('qt6base@' + v, when='@' + v)
        depends_on('qt6shadertools@' + v, when='@' + v)
```

sethrj reviewed (commented):
Love it so far! Most importantly, the justification for separate package is that these are now separate repos with CMake-based interdependencies -- a package model fits much better than trying to glom them together into a single package.

One change is that if we're going to start an "ecosystem" we should format the package directories as `qt-`. I think we can then upgrade the `qt` package so that it has a bunch of `when=` arguments to the variants, *except* those that enable the package choices (e.g. we could have `+opengl` simply add a `depends_on('qt6quick3d', when='@6:')`).

This would change the qt package to essentially be a "metapackage" for versions 6+, but allow the additional granularity of allowing the installation of just `qt-base` if needed. Then, as packages such as Geant4 add support for Qt6, they can change
```python
depends_on('qt')
```
to
```python
depends_on('qt', when='@:11.0)
depends_on('qt-base', when='@:11.1:)
```

wdconinc committed:
Qt6: fix style issues

sethrj commented on var/spack/repos/builtin/packages/qt6declarative/package.py:
```diff
@@ -3,9 +3,10 @@
-from spack import *
 import os
 
+from spack import *
```
These are apparently superfluous in current versions of spack -- you can delete the line entirely.

wdconinc committed:
[qt6] fix style issues, trailing spaces

wdconinc committed:
[qt6] rename to qt-* ecosystem; remove imports

wdconinc committed:
[qt6] rename dependencies; change version strings

wdconinc committed:
[qt6] list_urls

wdconinc committed:
[qt6] homepage links

wdconinc committed:
[qt6] missing closing quotes failed style check

sethrj commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,112 @@
+            self.define_from_variant('FEATURE_ssl', 'ssl'),
+            self.define('INPUT_openssl', 'linked'),
+            self.define_from_variant('FEATURE_widgets', 'widgets'),
+            self.define_from_variant('QT_BUILD_EXAMPLES', 'examples'),
+            self.define_from_variant('QT_BUILD_TESTS', 'tests'),
```
I don't think we want a `test` variant since it presumably doesn't change the installed binaries. Likewise, if the examples aren't installed they shouldn't get a variant.

wdconinc replied:
I'll change that to use `self.run_tests` on build.

sethrj commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,112 @@
+            self.define('FEATURE_system_pcre2', True),
+            self.define('FEATURE_system_png', True),
+            self.define('FEATURE_system_proxies', True),
+            self.define_from_variant('FEATURE_sql', 'sql'),
```
A little local helper function:
```python
     def define_feature(variant):
        return self.define_from_variant('FEATURE_' + variant, variant)
```

would help error-proof some of these I think. Likewise, I'd replace `self.define('FEATURE_system_' + x, True)` with a loop.

sethrj commented on var/spack/repos/builtin/packages/qt-declarative/package.py:
```diff
@@ -0,0 +1,49 @@
+    depends_on("pkgconfig", type='build')
+    depends_on("python", when='@5.7.0:', type='build')
+
+    versions = ['6.2.3']
+    for v in versions:
```
Best to use 'private' variables for class-level temporaries because they become part of the package. You don't want someone to add a `def versions(self)` to the package and conflict with it.

```suggestion
    _versions = ['6.2.3']
    for v in _versions:
```

wdconinc replied:
I'll propagate this change to the other places it is used too.

sethrj reviewed (commented):
@wdconinc any progress on this lately? If you think it's in a good enough state for me to test, I'll try it out on my M1 mac.

wdconinc committed:
qt-declarative: use private _versions

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>

wdconinc committed:
qt-quick3d, qt-quicktimeline, qt-shadertools: use private _versions

wdconinc committed:
qt-base: rework feature defines and use run_tests

wdconinc committed:
qt: new version 6.2.4

wdconinc committed:
flake8 whitespace before comma

wdconinc commented:
I have applied all outstanding changes, added new version 6.2.4, tested a full build on my system, rebased the branch to current develop, and set it to ready for review.

I have no idea what to expect on an M1 mac system, but please give it a try. I also need to spend some time doing a more complete check of the variant combinations.

sethrj commented:
Excellent! I'll try testing it on rhel7, macOS (Intel), and macOS (M1) today :)

sethrj commented:
@wdconinc : success on my rhel7 system with base level options after the libproxy fix (linked above)! I think I'll test a base-with-minimal configration, then merge this, and I'll open a separate PR for mac fixups (currently `libdrm` is listed as a requirement but it's incompatible with macOS).

sethrj commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,115 @@
+    depends_on("libtiff")
+    depends_on('libxkbcommon')
+    depends_on("libxml2")
+    depends_on('libxrender')
```
I think some of these dependencies may not be required when GUI is disabled. I'll post the CMake output when I build a "minimal" configuration so hopefully we can glean what CMake looks for.

sethrj commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,115 @@
+    variant('opengl', default=False, description='Build with OpenGL support.')
+    variant('shared', default=True, description='Build shared libraries.')
+    variant('sql', default=True, description='Build with SQL support.')
+    variant('ssl', default=True, description='Build with OpenSSL support.')
```
Qt internally calls  this option `openssl` so line `96` really should be `openssl`. Since many other packages name the equivalent variant `openssl` let's rename it:
```diff
diff --git a/var/spack/repos/builtin/packages/qt-base/package.py b/var/spack/repos/builtin/packages/qt-base/package.py
index bcf0727cee..8361fcc0f9 100644
--- a/var/spack/repos/builtin/packages/qt-base/package.py
+++ b/var/spack/repos/builtin/packages/qt-base/package.py
@@ -33,7 +33,7 @@ class QtBase(CMakePackage):
     variant('opengl', default=False, description='Build with OpenGL support.')
     variant('shared', default=True, description='Build shared libraries.')
     variant('sql', default=True, description='Build with SQL support.')
-    variant('ssl', default=True, description='Build with OpenSSL support.')
+    variant('openssl', default=True, description='Build with OpenSSL support.')
     variant('widgets', default=True, description='Build with widgets.')

     depends_on('cmake@3.16:', type='build')
@@ -59,7 +59,7 @@ class QtBase(CMakePackage):
     depends_on('libxkbcommon')
     depends_on("libxml2")
     depends_on('libxrender')
-    depends_on('openssl', when='+ssl')
+    depends_on('openssl', when='+openssl')
     depends_on('pcre2+multibyte')
     depends_on("sqlite", when='+sql')
     depends_on("zlib")
@@ -93,7 +93,7 @@ def define_feature(variant):
             define_feature('dbus'),
             define_feature('framework'),
             define_feature('gui'),
-            define_feature('ssl'),
+            define_feature('openssl'),
             define_feature('widgets'),
             define_feature('sql'),
         ]
```
Currently `~ssl` results in an error:
```
2 errors found in build log:
     220    -- Performing Test HAVE_dtls
     221    -- Performing Test HAVE_dtls - Success
     222    -- Performing Test HAVE_ocsp
     223    -- Performing Test HAVE_ocsp - Success
     224    -- Performing Test HAVE_networklistmanager
     225    -- Performing Test HAVE_networklistmanager - Failed
  >> 226    CMake Error at cmake/QtBuildInformation.cmake:388 (message):
     227      Feature "openssl": Forcing to "ON" breaks its condition:
     228
     229          QT_FEATURE_openssl_runtime OR QT_FEATURE_openssl_linked
     230
     231      Condition values dump:
     232

     ...

     414        QT_FEATURE_openssl_runtime OR QT_FEATURE_openssl_linked
     415    Condition values dump:
     416        QT_FEATURE_openssl_runtime = "OFF"
     417        QT_FEATURE_openssl_linked = "OFF"
     418
     419
  >> 420    CMake Error at cmake/QtBuildInformation.cmake:80 (message):
     421      Check the configuration messages for an error that has occurred.
     422    Call Stack (most recent call first):
     423      cmake/QtBuildInformation.cmake:11 (qt_configure_print_summary)
     424      cmake/QtBuildInternals/QtBuildInternalsConfig.cmake:472 (qt_print_feature_summary)
     425      CMakeLists.txt:192 (qt_build_repo_end)
     426
```

sethrj replied:
Actually, with openssl off, qt still looks for `libcrypto` but doesn't use it. Instead it looks for and links to a system GSSAPI package, which we don't want...

Maybe we need a variant `+network` that defaults to true, and require openssl if enabled?

wdconinc replied:
I think Qt has a mixture of [FEATURE_ssl](https://github.com/qt/qtbase/blob/dev/src/network/CMakeLists.txt#L322) and [FEATURE_openssl](https://github.com/qt/qtbase/blob/dev/src/plugins/tls/CMakeLists.txt#L5)...

sethrj replied:
Oh dear. Do you have a preference on a network component variant versus twiddling with the ssl dependency? I generally think package variants should be defined by *capabilities* they enable rather than by *dependencies* they require, so `network` would be more appropriate.

wdconinc replied:
I think a `network` variant would make sense, in particular with since it's a fairly toplevel flag too, https://github.com/qt/qtbase/blob/27efadec226cabd14f417c9f44f86557ea88ec4c/src/CMakeLists.txt#L8. I think there are still plenty of places where the Qt6 docs are following the older configure options of Qt5 even when maybe the build system is more structured now.

sethrj commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,115 @@
+        # FEATURE_system_* arguments: use system where possible
+        for x in ['doubleconversion', 'freetype', 'harfbuzz', 'jpeg',
+                  'pcre2', 'png', 'proxies', 'sqlite', 'zlib']:
+            args.append(self.define('FEATURE_system_' + x, True))
```
Installing with everything off:
```
CMake Warning:
  Manually-specified variables were not used by the project:

    FEATURE_accessibility
    FEATURE_system_freetype
    FEATURE_system_harfbuzz
    FEATURE_system_jpeg
    FEATURE_system_png
    FEATURE_system_sqlite
    FEATURE_system_textmarkdownreader
    INPUT_opengl
```

suggesting those dependencies aren't needed.

wdconinc replied:
Maybe something like this? (untested)
```python
        # FEATURE_system_* arguments: use system where possible
        def set_feature_system_flags(feature_system, value):    
            for flag, variant in feature_system:
                if variant is None or self.satisfies(variant):
                    args.append(self.define('FEATURE_system_' + flag, value))   
        feature_system_true = {
            'doubleconversion': None,
            'freetype': '+gui',
            'harfbuzz': '+gui',
            'jpeg': '+gui',
            'pcre2': None,
            'png': '+gui',
            'proxies': None,
            'sqlite': '+sql',
            'zlib': None,
        }
        feature_system_false = {
            'libb2': None,
            'textmarkdownreader': None,
        }
        set_feature_system_flags(feature_system_true, True)
        set_feature_system_flags(feature_system_false, False)
```
Could possibly be even more generic to reuse for other lists that are variant dependent.

sethrj replied:
Maybe—the key thing is that its output gives us a list of `depends_on` that should be conditional based on gui/widgets.

sethrj commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,115 @@
+    variant('dbus', default=False, description='Build with D-Bus support.')
+    variant('examples', default=False, description='Build examples.')
+    variant('framework', default=False, description='Build as a macOS Framework package.')
+    variant('gtk', default=False, description='Build with gtkplus.')
```
This variant might only be applicable when `+widgets`? It also is inapplicable to macos.

sethrj commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,115 @@
+    variant('framework', default=False, description='Build as a macOS Framework package.')
+    variant('gtk', default=False, description='Build with gtkplus.')
+    variant('gui', default=True, description='Build the Qt GUI module and dependencies.')
+    variant('opengl', default=False, description='Build with OpenGL support.')
```
OpenGL is only applicable when `+gui`.

wdconinc replied:
```suggestion
    variant('opengl', default=False, description='Build with OpenGL support.', when='+gui')
```

sethrj reviewed (changes requested):
I've tried a few combinations on rhel7 and dug into the CMakelists to figure out what the variant interdependencies are. The options in the base package I assume are taken from the old qt package, and they don't quite reflect the major feature options in the main CMakeLists:
- concurrent (default on)
- sql ✅ 
- network (default on)
- xml (default on)
- dbus ✅ 
- gui ✅ 
- gui/opengl  ✅ 
- gui/widgets  ✅ 
- testlib (default on)
- printsupport (default on)

wdconinc committed:
qt-base: variant opengl when +gui

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>

sethrj committed:
Update features based on gui usage

sethrj committed:
Update dependencies, cmake args, mac support

sethrj committed:
Update features based on linux

sethrj committed:
More updates

wdconinc committed:
qt6: rebase and apply new black style

wdconinc committed:
qt6: apply style isort fixes

wdconinc committed:
qt6: new version 6.3.0 and 6.3.1

wdconinc commented:
Added newer versions 6.3.0 and 6.3.1.

Full change sets:
| versions | qt-base | qt-declarative | qt-quick3d | qt-quicktimeline | qt-shadertools |
|---|---|---|---|---|---|
| 6.2.4 to 6.3.0 | [diff](https://github.com/qt/qtbase/compare/v6.2.4...v6.3.0) | [diff](https://github.com/qt/qtdeclarative/compare/6.2.4...v6.3.0) | [diff](https://github.com/qt/qtquick3d/compare/6.2.4...v6.3.0) | [diff](https://github.com/qt/qtquicktimeline/compare/6.2.4...v6.3.0) | [diff](https://github.com/qt/qtshadertools/compare/6.2.4...v6.3.0) |
| 6.3.0 to 6.3.1 | [diff](https://github.com/qt/qtbase/compare/v6.3.1...v6.3.1) | [diff](https://github.com/qt/qtdeclarative/compare/6.3.0...v6.3.1) | [diff](https://github.com/qt/qtquick3d/compare/6.3.0...v6.3.1) | [diff](https://github.com/qt/qtquicktimeline/compare/6.3.0...v6.3.1) | [diff](https://github.com/qt/qtshadertools/compare/6.3.0...v6.3.1) |

No major build system changes from what I could tell, but still need to run test builds.

wdconinc commented:
Version 6.3.1 builds fine on my system:
```console
5clmspe qt-base@6.3.1+accessibility~dbus~examples~framework~gtk+gui~ipo~opengl+shared+sql+ssl+widgets build_type=Release dev_path=/data/develop/qt-base-6.3.1
7jnwd3y qt-declarative@6.3.1~ipo build_type=Release dev_path=/data/develop/qt-declarative-6.3.1
6b5qjjd qt-quick3d@6.3.1~ipo build_type=Release dev_path=/data/develop/qt-quick3d-6.3.1
c6v5qio qt-quicktimeline@6.3.1~ipo build_type=RelWithDebInfo dev_path=/data/develop/qt-quicktimeline-6.3.1
ztdh4st qt-shadertools@6.3.1~ipo build_type=Release dev_path=/data/develop/qt-shadertools-6.3.1
```

Now I can get back to the discussion we were having on various combinations of variants...

wdconinc committed:
qt6: add 6.3.0 and 6.3.1 to versions list

wdconinc commented on var/spack/repos/builtin/packages/qt-declarative/package.py:
```diff
@@ -0,0 +1,58 @@
+    depends_on("pkgconfig", type="build")
+    depends_on("python", when="@5.7.0:", type="build")
+
+    _versions = ["6.3.1", "6.3.0", "6.2.4", "6.2.3"]
+    for v in _versions:
+        depends_on("qt-base@" + v, when="@" + v)
+        depends_on("qt-shadertools@" + v, when="@" + v)
```
This anti-pattern of storing the version list twice seems like it should be possible to avoid...

adamjstewart replied:
See https://github.com/spack/spack/pull/14002 for some preliminary work to make this simpler.

wdconinc replied:
What is the timeline on that pr getting finalized?

adamjstewart replied:
No timeline ☹️

adamjstewart commented:
I'm trying to add a `py-pyqt6` package as a possible matplotlib backend and I'm wondering if this PR is still a WIP.

wdconinc committed:
Merge branch 'develop' into new-package-qt6

wdconinc commented:
> I'm wondering if this PR is still a WIP.

Yes, and no. I think what's been mostly missing is a downstream package to test this with. Until then it is mostly 'just' a library...

I had personally been waiting to get this working with geant4, but qt6 support there didn't appear yet in a spack release.

I have been meaning to dissect the beta of the [next release](https://gitlab.cern.ch/geant4/geant4/-/tags/v11.1.0.beta) but the geant4 release model to their public repositories is simply not conducive to this kind of exercise (single squash commit with all changes for a new release, no insight into current development).

The status is that this compiles and installs fine, but we are not quite sure what the best way is to structure variants for feature inclusion. See discussions.

As far as I'm concerned, I'd say we merge this (it compiles and is functional to the extent it can be tested without downstream packages) but that we are somehow explicit about the fact that downstream packages may need to expect changes in variants etc in the short term.

adamjstewart commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,142 @@
+    depends_on("zstd")
+
+    def patch(self):
+        import os
+        import shutil
```
Can you move all imports to the top of the file?

adamjstewart commented on var/spack/repos/builtin/packages/qt-base/package.py:
```diff
@@ -0,0 +1,142 @@
+        import os
+        import shutil
+
+        vendor_dir = join_path(self.stage.source_path, "src/3rdparty")
```
```suggestion
        vendor_dir = join_path(self.stage.source_path, "src", "3rdparty")
```

adamjstewart commented on var/spack/repos/builtin/packages/qt-shadertools/package.py:
```diff
@@ -0,0 +1,59 @@
+                    if dep not in vendor_deps_to_keep:
+                        shutil.rmtree(dep)
+
+    def cmake_args(self):
+        args = []
+        return args
```
This can be removed if you aren't going to use it

adamjstewart reviewed (commented):
Will try building on macOS now

wdconinc committed:
qt6: multi-argument join_path

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

wdconinc committed:
qt-base: fix isort

wdconinc committed:
qt-shadertools: no cmake_args needed

adamjstewart commented on var/spack/repos/builtin/packages/qt-declarative/package.py:
```diff
@@ -0,0 +1,58 @@
+
+    def patch(self):
+        import os
+        import shutil
```
Same here

adamjstewart commented on var/spack/repos/builtin/packages/qt-declarative/package.py:
```diff
@@ -0,0 +1,58 @@
+        import os
+        import shutil
+
+        vendor_dir = join_path(self.stage.source_path, "src/3rdparty")
```
Same here

adamjstewart commented on var/spack/repos/builtin/packages/qt-quick3d/package.py:
```diff
@@ -0,0 +1,66 @@
+
+    def patch(self):
+        import os
+        import shutil
```
Same here

adamjstewart commented on var/spack/repos/builtin/packages/qt-quick3d/package.py:
```diff
@@ -0,0 +1,66 @@
+        import os
+        import shutil
+
+        vendor_dir = join_path(self.stage.source_path, "src/3rdparty")
```
Same here

adamjstewart commented on var/spack/repos/builtin/packages/qt-declarative/package.py:
```diff
@@ -0,0 +1,58 @@
+
+    def cmake_args(self):
+        args = []
+        return args
```
Same here

adamjstewart commented on var/spack/repos/builtin/packages/qt-shadertools/package.py:
```diff
@@ -0,0 +1,55 @@
+
+    def patch(self):
+        import os
+        import shutil
```
Same here

adamjstewart commented on var/spack/repos/builtin/packages/qt-shadertools/package.py:
```diff
@@ -0,0 +1,55 @@
+        import os
+        import shutil
+
+        vendor_dir = join_path(self.stage.source_path, "src/3rdparty")
```
Same here

wdconinc committed:
qt-declarative: imports up front

wdconinc committed:
qt-quick3d: fix import

wdconinc committed:
qt-declarative: remove useless cmake_args

wdconinc committed:
qt-shadertools: imports and join_path fixes

wdconinc committed:
qt-quick3d: join_path fixes

wdconinc committed:
qt-declarative: join_path fixes

sethrj commented:
Sorry @wdconinc and @adamjstewart for dropping the ball on this. Early may I spent a few hours poking around with this, digging through the qt CMake code to find the workable options and see what happened on linux and/or mac.

https://github.com/sethrj/spack/compare/792d498beb13469b5930cd298de253cad09235ee..eff0c63d286bf4e57f97a0fcbfe8dfc16f98db98

is the point that I got to (here's the branch: https://github.com/sethrj/spack/tree/new-package-qt6) but I'm sure it's terribly out of date now and full of conflicts. However, I know the recipe there is much closer to the way that qt (at least version 6.3.0) used dependencies with the cmake scripts.

wdconinc commented:
I may try to cherry-pick those into this PR, if that's ok with you.

sethrj commented:
@wdconinc absolutely! mainly I focused on trimming down the dependency lists on the base to match the actual usage based on the CMakeCache and CMakeLists in Qt. If I recall correctly one of the blockers was that I wasn't able to disable ssl; and some of the variants are unsupported/low-level switches in the cmake that might change from one minor version to the next.

wdconinc commented:
@spackbot fix style

wdconinc committed:
qt-base: fix style

wdconinc commented:
With these latest cherry-picks, I'm getting the following configure phase failure (for `qt-base +gui`):
```console
1 error found in build log:
     43    
     44    
     45    Note: When linking against OpenSSL, you can override the default library names through OPENSSL_LIBS. For example: OPENSSL_LIBS='-L/opt/ssl/lib -lssl -lcrypto' ./configure -openssl-linked
     46    Note: Disabling X11 Accessibility Bridge: D-Bus or AT-SPI is missing.
     47    
     48    -- Configuring done
  >> 49    CMake Error at src/corelib/Qt6CoreMacros.cmake:2440 (add_library):
     50      Cannot find source file:
     51    
     52        /home/wdconinc/.spack/stage/spack-stage-qt-base-6.3.1-2qiyhdgndf3rxonbo7kc2s6c6w3cp44l/spack-src/src/3rdparty/xcb/libxcb/xinput.c
     53    Call Stack (most recent call first):
     54      cmake/QtModuleHelpers.cmake:128 (_qt_internal_add_library)
     55      src/plugins/platforms/xcb/CMakeLists.txt:7 (qt_internal_add_module)
```

wdconinc commented:
Related to above but not in the log blurb that spack shows:
```console
CMake Warning (dev) at cmake/3rdparty/extra-cmake-modules/find-modules/FindXCB.cmake:180 (message):
  XINPUT from XCB was requested: this is EXPERIMENTAL and is likely to
  unavailable on many systems!
Call Stack (most recent call first):
  cmake/QtFindPackageHelpers.cmake:130 (find_package)
  src/gui/configure.cmake:123 (qt_find_package)
  cmake/QtFeature.cmake:573 (include)
  src/CMakeLists.txt:12 (qt_feature_evaluate_features)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

This requires an additional `+gui` flag for `FEATURE_system` with `xcb_xinput`:
```diff
diff --git a/var/spack/repos/builtin/packages/qt-base/package.py b/var/spack/repos/builtin/packages/qt-base/package.py
index c273525411..c890b296de 100644
--- a/var/spack/repos/builtin/packages/qt-base/package.py
+++ b/var/spack/repos/builtin/packages/qt-base/package.py
@@ -176,6 +176,7 @@ def define_feature(key, variant=None):
                 ("freetype", True),
                 ("harfbuzz", True),
                 ("textmarkdownreader", False),
+                ("xcb_xinput", True),
             ]
         if "+network" in spec:
             sys_features += [
```
and likely a version requirement on `libxcb` to ensure it actually has that (recently added) `xinput` support. It was added in `libxcb@1.12`, but enabled by default in `libxcb@1.13`, so we probably don't build it in spack until `libxcb@1.13`.

wdconinc commented:
@sethrj Was the `archive_files` for debugging only? Or did you want to keep it around?

wdconinc commented:
Mainly asking about `archive_files` because:
```diff
diff --git a/var/spack/repos/builtin/packages/qt-base/package.py b/var/spack/repos/builtin/packages/qt-base/package.py
index c273525411..297e71e9e2 100644
--- a/var/spack/repos/builtin/packages/qt-base/package.py
+++ b/var/spack/repos/builtin/packages/qt-base/package.py
@@ -84,7 +84,7 @@ class QtBase(CMakePackage):
     def archive_files(self):
         """Save both the CMakeCache and the config summary."""
         return [
-            join(self.build_directory, filename)
+            join_path(self.build_directory, filename)
             for filename in ["CMakeCache.txt", "config.summary"]
         ]
 
 ```

sethrj commented:
Yeah, the config output is a good extra diagnostic to keep around in addition to the cmake output

wdconinc committed:
qt-base: archive_files join_path

wdconinc committed:
qt-base: new version 6.3.2

wdconinc committed:
qt-{declarative,quick3d,quicktimeline,shadertools}@6.3.2

wdconinc committed:
qt-base: require libxcb@1.13: and use system xcb_xinput when on linux

adamjstewart commented:
Just tried building on macOS:
```
2 errors found in build log:
     11     -- Detecting CXX compile features - done
     12     -- Detecting C compiler ABI info
     13     -- Detecting C compiler ABI info - done
     14     -- Check for working C compiler: /Users/ajstewart/spack/lib/spack/env/clang/clang - skipped
     15     -- Detecting C compile features
     16     -- Detecting C compile features - done
  >> 17     CMake Error at cmake/QtBuildInformation.cmake:472 (message):
     18       Feature "libproxy": Forcing to "ON" breaks its condition:
     19     
     20           Libproxy_FOUND
     21     
     22       Condition values dump:
     23     

     ...

     220    ERROR: Feature "libproxy": Forcing to "ON" breaks its condition:
     221        Libproxy_FOUND
     222    Condition values dump:
     223        Libproxy_FOUND = "0"
     224    
     225    
  >> 226    CMake Error at cmake/QtBuildInformation.cmake:163 (message):
     227      Check the configuration messages for an error that has occurred.
     228    Call Stack (most recent call first):
     229      cmake/QtBuildInformation.cmake:20 (qt_configure_print_summary)
     230      cmake/QtBuildInternals/QtBuildInternalsConfig.cmake:501 (qt_print_feature_summary)
     231      CMakeLists.txt:193 (qt_build_repo_end)
     232    
```

wdconinc commented:
I updated the `qt-base` package with a new version, and added the requirement for linux to use `libxcb@1.13:` which has xinput, and then use that system_xcb_xinput. That gets me a successful build on gcc-12.
```console
==> qt-base: Successfully installed qt-base-6.3.2-v4psplcdpztf7jknalairnbbxvmqd5kk
```

I unfortunately don't have a mac (and now even github doesn't have macos-1015 targets anymore).

sethrj commented:
I'm personally fine with getting Linux to work for now and fixing the macOS build in a follow on commit...

wdconinc commented:
I assume there is no way to convince the gitlab pipelines to build for mac?

adamjstewart commented:
We have a macOS pipeline, but it's currently broken, and idk if anyone is actively working on fixing it, so it's optional for now.

sethrj committed:
Merge branch 'develop' into new-package-qt6

adamjstewart commented:
Any updates on this PR?

sethrj reviewed (approved):
Sorry I let this slide for so long! The main variants look good so let's merge it and fix any issues in subsequent pulls. Thanks again @wdconinc for this addition.
